### DISCUSSION TITLE
Get Null space of an Operator

### DISCUSSION THREAD

**Discussion by mfeuerle** [pymor/pymor#1532] [Q&A]
Is there an easy way to get any informations  about the null space / kernel of a (linear) `Operator`? Either a basis of the null space, or a projection onto the null space or basically any information which makes the null space (or a approximation of the null space) accessible?

**Comment by pmli** [comment]
This sounds like a difficult problem for a generic linear `Operator`. If the dimension of the null space is too large, computing its basis might be too expensive. Do you expect that the dimension is small? In that case, `pymor.algorithms.eig.eig` might work.

**Reply by mfeuerle** [reply]
I would assume that it could be large. The dimension of the `Operator` itself right now is quite small, lets say < 5.000, but in the future it will probably grow to 6 digits. I would assume, that the dimension of the null space could be 1/4 of the `Operator` dimension. In the current MATLAB implementation the `null()` method is used, which internally uses `svd`. So the `pymor.algorithms.svd_va` module could be worth a shot. The longer i thing about it, the more it seems that this is not really a `pymor` related problem but a general numerical problem. As far as i am aware, nearly all algorithms are based on either `eig` or `svd` and there are actually no really efficient algorithms for calculating a basis (or just a projection) of the null space of a (sparse) matrix...

**Reply by pmli** [reply]
Ok, 1/4 of the full dimension sounds too big even to store the basis. If it were bigger, i.e., if the `Operator` was of low numerical rank, one of the methods from `pymor.algorithms.randrangefinder` could be used to approximate the range, which could then be used to define a projection onto the null space.
I don't have other ideas to represent the null space. A problem-specific approach might be necessary.

**Reply by mfeuerle** [reply]
That sounds link an interesting idea. Do you already have a (vague) idea in mind, how the range could be used to define a projection onto the null space? The only connection i am aware of right now is `dim(Operator) = dim(null space) + dim(range)`, but that's only a connection between the dimensions.

**Reply by pmli** [reply]
One trick is to use that the null space is the orthogonal complement of the adjoint operator's range. So, if `V` is an ONB of the adjoint operator's range, then `I - V^H V` is a projector onto the null space.

**Reply by sdrave** [reply]
So this approach would give you a projection onto a subspace which is guaranteed to contain the kernel. However, in most cases it will be a strictly larger space.

**Reply by mfeuerle** [reply]
Okey, that is an approach i will keep in mind, maybe as a default fallback option, if no additional user provided informations are available. Thank you :-)

**Comment by drittelhacker** [comment]
This is the main problem why large scale DAE reduction by anything like balanced truncation is such a tough problem. If you don't know anything about your problem that allows you to derive this kind of information, e.g. from the structure of the problem, there is little to no hope  to compute this reliablyand efficiently. For moderate dimensions one can do this, e.g. via the matrix sign function iteration, but since that computes a matrix inverse in each step, it quickly becomes infeasible when dimensions grow. So indeed it a general rather than a pyMOR-related problem.

**Reply by mfeuerle** [reply]
Thank you. I suppose, that i will then implement a general default option for small and medium scaled problems. But if that reaches its limits there is no way around than manually exploiting the problem structure to progress further.
